File: src/Library/Tests/TestCppCXUWP/PluralConverter.cpp
#include "pch.h"
#include "PluralConverter.h"
#include "StringFormat.h"

using namespace Windows::ApplicationModel::Resources;
using namespace Huyn::ReswPlusLib;
using namespace Platform;
using namespace Windows::UI::Xaml::Interop;

namespace TestCppCXUWP {
	namespace Converters {
		PluralConverter::PluralConverter()
		{
		}

		PluralConverter::~PluralConverter()
		{
		}

		Object^ PluralConverter::Convert(Object^ value, TypeName targetType, Object^ parameter, String^ language)
		{
			auto key = static_cast<Platform::String^>(parameter);
			if (key == nullptr || key->IsEmpty())
			{
				return L"";
			}

			auto number = safe_cast<double>(value);

			auto resource = ResourceLoader::GetForCurrentView();
			auto pluralFormat = ResourceLoaderExtension::GetPlural(ResourceLoader::GetForCurrentView(), "RunDistance", number);
			wchar_t * str;
			if (vasprintf(&str, pluralFormat->Data(), number) >= 0)
			{
				auto res = ref new Platform::String(str);
				delete[] str;
				return res;
			}
			return L"";
		}

		Object^ PluralConverter::ConvertBack(Object^ value, TypeName targetType, Object^ parameter, String^ language)
		{
			return nullptr;
		}
	}
}

File: src/Library/Tests/TestCSharpUWP/Converters/PluralConverter.cs
using ReswPlusLib;
using System;
using System.Globalization;
using System.Resources;
using Windows.ApplicationModel.Resources;
using Windows.UI.Xaml.Data;

namespace TestCSharpUWP.Converters
{
    public sealed class PluralConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, string language)
        {
            var culture = !string.IsNullOrWhiteSpace(language) ? new CultureInfo(language) : null;
            var key = parameter as string;
            if (string.IsNullOrWhiteSpace(key))
            {
                return string.Empty;
            }
            var number = System.Convert.ToDouble(value);

            var resource = ResourceLoader.GetForCurrentView();
            var pluralFormat = resource.GetPlural(key, number);

            return string.Format(culture, pluralFormat, number);
        }

        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            return null;
        }
    }
}
